nemo/collections/nlp/modules/common/megatron/position_embedding/alibi_relative_position_embedding.py

```diff
@@ -42,20 +42,31 @@ def build_slopes(num_attention_heads, num_attention_heads_alibi):
     """
     Builds a slopes tensor.
     """
-    slopes = torch.Tensor(
-        get_slopes(num_attention_heads_alibi) + [0] * (num_attention_heads - num_attention_heads_alibi)
-    ).cuda()
-    return slopes.unsqueeze(-1).unsqueeze(-1)
+    slopes = (
+        torch.Tensor(get_slopes(num_attention_heads_alibi) + [0] * (num_attention_heads - num_attention_heads_alibi))
+        .unsqueeze(-1)
+        .unsqueeze(-1)
+    )
 
+    if torch.cuda.is_available():
+        slopes = slopes.to(torch.cuda.current_device())
 
-def build_relative_position(query_length, key_length, num_attention_heads):
-    context_position = torch.arange(query_length)[:, None].cuda()
-    memory_position = torch.arange(key_length)[None, :].cuda()
-    # shape (query_length, key_length, num_heads)
-    relative_position = memory_position - context_position
+    return slopes
+
+
+def build_relative_position(max_seq_len, full=True):
+    """
+    full=True:  shape (max_seq_len, max_seq_len)
+    full=False: shape (max_seq_len)
+    """
+    relative_position = torch.arange(1 - max_seq_len, 1)[None, :].mul(-1)  # (1, max_seq_len)
+
+    if full:
+        memory_position = torch.arange(1 - max_seq_len, 1)[:, None].mul(-1)
+        relative_position = torch.abs(memory_position - relative_position)  # (max_seq_len, max_seq_len)
 
-    # shape (num_attention_heads, max_seq_len, max_seq_len)
-    relative_position = torch.abs(relative_position).unsqueeze(0).expand(num_attention_heads, -1, -1)
+    if torch.cuda.is_available():
+        relative_position = relative_position.to(torch.cuda.current_device())
```

Comment:
This seems strange why are you checking for cuda here and moving the tensor manually?

Reply:
The original code use `.cuda()` . However @trias702 suggest we need support CPU-only one.

Reply:
Yes, I saw that we have explicit support for CPUAccelerator inside MegatronGPTModel here: https://github.com/NVIDIA/NeMo/blob/ebfcef7902c9232da04ecf90f505c3c996e7e059/nemo/collections/nlp/models/language_modeling/megatron_gpt_model.py#L215

That would cause a device incompatibility to be thrown if you're doing a manual .cuda() call somewhere in the downstream Megatron classes, because your model at instantiation expects everything to be on CPU.